File: TpPoo_cpp/tp2_ex3/VoitureAutonome.cpp
#include <iostream>

class VoitureAutonome {
private:
    double position;
    double vitesse;
    double distanceParcourue;

public:
    VoitureAutonome(double pos, double vit) : position(pos), vitesse(vit), distanceParcourue(0.0) {}

    friend bool operator ==(const VoitureAutonome &v1, const VoitureAutonome &v2);
    friend bool operator !=(const VoitureAutonome &v1, const VoitureAutonome &v2);
    friend bool operator <(const VoitureAutonome &v1, const VoitureAutonome &v2);
    friend bool operator >(const VoitureAutonome &v1, const VoitureAutonome &v2);
    friend bool operator <=(const VoitureAutonome &v1, const VoitureAutonome &v2);
    friend bool operator >=(const VoitureAutonome &v1, const VoitureAutonome &v2);
    friend VoitureAutonome operator +(const VoitureAutonome &v1, const VoitureAutonome &v2);
    friend VoitureAutonome operator ++(VoitureAutonome &v1);
    friend VoitureAutonome operator -(const VoitureAutonome &v1, const VoitureAutonome &v2);
    friend VoitureAutonome operator *(const VoitureAutonome &v1, const VoitureAutonome &v2);
    friend VoitureAutonome operator /(const VoitureAutonome &v1, const VoitureAutonome &v2);

    friend std::ostream& operator<<(std::ostream& os, const VoitureAutonome &car);
};


bool operator ==(const VoitureAutonome &v1, const VoitureAutonome &v2) {
    return v1.position == v2.position && v1.vitesse == v2.vitesse;
}

bool operator !=(const VoitureAutonome &v1, const VoitureAutonome &v2) {
    return !(v1 == v2);
}

bool operator <(const VoitureAutonome &v1, const VoitureAutonome &v2) {
    return v1.position < v2.position;
}

bool operator >(const VoitureAutonome &v1, const VoitureAutonome &v2) {
    return v1.position > v2.position;
}

bool operator <=(const VoitureAutonome &v1, const VoitureAutonome &v2) {
    return !(v1 > v2);
}

bool operator >=(const VoitureAutonome &v1, const VoitureAutonome &v2) {
    return !(v1 < v2);
}

VoitureAutonome operator +(const VoitureAutonome &v1, const VoitureAutonome &v2) {
    VoitureAutonome result = v1;
    result.distanceParcourue += v2.distanceParcourue;
    return result;
}

VoitureAutonome operator ++(VoitureAutonome &v1) {
    v1.vitesse++;
    return v1;
}

VoitureAutonome operator -(const VoitureAutonome &v1, const VoitureAutonome &v2) {
    VoitureAutonome result = v1;
    result.distanceParcourue -= v2.distanceParcourue;
    return result;
}

VoitureAutonome operator *(const VoitureAutonome &v1, const VoitureAutonome &v2) {
    VoitureAutonome result = v1;
    return result;
}

VoitureAutonome operator /(const VoitureAutonome &v1, const VoitureAutonome &v2) {
    VoitureAutonome result = v1;
    return result;
}

std::ostream& operator<<(std::ostream& os, const VoitureAutonome &car) {
    os << "Position: " << car.position << ", Vitesse: " << car.vitesse << ", Distance Parcourue: " << car.distanceParcourue;
    return os;
}

int main() {
    VoitureAutonome car1(0.0, 50.0);
    VoitureAutonome car2(10.0, 60.0);

    if (car1 == car2) {
        std::cout << "Les voitures sont identiques." << std::endl;
    } else {
        std::cout << "Les voitures sont différentes." << std::endl;
    }
    VoitureAutonome totalDistance = car1 + car2;
    std::cout << "Distance totale parcourue : " << totalDistance << std::endl;

    ++car1;
    std::cout << "Nouvelle vitesse de car1 : " << car1 << std::endl;

    VoitureAutonome remainingDistance = car1 - car2;
    std::cout << "Distance restante : " << remainingDistance << std::endl;

    VoitureAutonome resultMultiply = car1 * car2;
    VoitureAutonome resultDivide = car1 / car2;

    std::cout << "Résultat de la multiplication : " << resultMultiply << std::endl;
    std::cout << "Résultat de la division : " << resultDivide << std::endl;

    return 0;
}

File: Poo_java/tp2/tp_ex3/src/VoitureAutonome.java
public class VoitureAutonome {
    private double position;
    private double vitesse;
    private double distanceParcourue;

    public VoitureAutonome(double pos, double vit) {
        this.position = pos;
        this.vitesse = vit;
        this.distanceParcourue = 0.0;
    }

    public boolean equals(VoitureAutonome other) {
        return this.position == other.position && this.vitesse == other.vitesse;
    }

    public boolean notEquals(VoitureAutonome other) {
        return !this.equals(other);
    }

    public boolean lessThan(VoitureAutonome other) {
        return this.position < other.position;
    }

    public boolean greaterThan(VoitureAutonome other) {
        return this.position > other.position;
    }

    public boolean lessThanOrEqual(VoitureAutonome other) {
        return !this.greaterThan(other);
    }

    public boolean greaterThanOrEqual(VoitureAutonome other) {
        return !this.lessThan(other);
    }

    public VoitureAutonome add(VoitureAutonome other) {
        VoitureAutonome result = new VoitureAutonome(this.position, this.vitesse);
        result.distanceParcourue += other.distanceParcourue;
        return result;
    }

    public VoitureAutonome increment() {
        this.vitesse++;
        return this;
    }

    public VoitureAutonome subtract(VoitureAutonome other) {
        VoitureAutonome result = new VoitureAutonome(this.position, this.vitesse);
        result.distanceParcourue -= other.distanceParcourue;
        return result;
    }

    public VoitureAutonome multiply(VoitureAutonome other) {
        VoitureAutonome result = new VoitureAutonome(this.position, this.vitesse);
        return result;
    }

    public VoitureAutonome divide(VoitureAutonome other) {
        VoitureAutonome result = new VoitureAutonome(this.position, this.vitesse);
        return result;
    }

    @Override
    public String toString() {
        return "Position: " + this.position + ", Vitesse: " + this.vitesse + ", Distance Parcourue: " + this.distanceParcourue;
    }

}